File: concepts/01 Array/04 Array and Pointers/04.c
// Array and Pointers

#include <stdio.h>

int main()
{
    int a[] = {6, 2, 1, 5, 3};
    int *p = a;
    printf("%p %p\n", a, p);
    p++;
    printf("%p %d\n", p, *p);
    int i = 2;
    printf("%d %d\n", a[i], i[a]);
    printf("%p %p\n", a+1, &a + 1);
    printf("%d %d", *(a+1), *a + 1);
}

File: concepts/01 Array/04 Array and Pointers/04.zig
// Array and Pointers

const std = @import("std");

// Zig doesn't follow same array pointer concept as C.
// * is format specifer for address in zig
// For pointer arithmetic and slicing use multi item pointer [*]T insted of *T.
// const a: [?]u32 = [_]u32{6, 2, 1, 5, 3};
// &a = [5]100 and a != &a
// var p: [*]u32 = &a
// We can't derefrence this multi item pointer like p.*
// Instead it would be used same as array name.
// p[1] = 2

pub fn main() !void {
    const a = [_]u8{ 6, 2, 1, 5, 3 };
    var p: [*]const u8 = &a;
    p += 1;
    std.debug.print("{any} {any}\n", .{ a[0], p[0] });
}
